### PR TITLE
Add icrc1Transfer method to ICP ledger canister

### PR DESCRIPTION
**Pull request by dskloetd**
# Motivation

We want to transfer ICP using the ICRC-1 API because then we don't have to compute account identifiers in the client.
Passing the ICP ledger canister ID to the ICRC-1 API works but the ICP ledger canister has its own ICRC-1 interface so calling it directly on the ICP ledger API is cleaner.

*Note*: I started with a commit that duplicates the non-ICRC-1 code. That way you can see how the ICRC-1 code differs from the existing code by reviewing the second commit.

# Changes

1. Added `icrc1Transfer` canister method.
2. Added `toIcrc1TransferRawRequest` request converter.
3. Added `mapIcrc1TransferError` error converter.
4. Added `Icrc1TransferRequest` request type.

# Tests

Copied the existing tests for `transfer` and applied them to `icrc1Transfer` as far as they made sense. The new method has less custom code for hardware wallets so some of those test have not been duplicated.

Tested manually in NNS dapp by using the new method when making ICRC-1 transfers and then testing an ICRC-1 transfer.

### CODE REVIEW

**Review by lmuntaner — APPROVED** [review]
For HW, we will need to check the version of the Ledger App to use this method or the old one.

File: packages/nns/src/ledger.canister.ts
Comment (peterpeterparker) [inline packages/nns/src/ledger.canister.ts:141]: prefer `isNullish` for consistency

File: packages/nns/src/ledger.canister.ts
Comment (peterpeterparker) [inline packages/nns/src/ledger.canister.ts:138]: nth: destruct `{fee, ...rest}`

File: packages/nns/src/ledger.canister.ts
Comment (peterpeterparker) [inline packages/nns/src/ledger.canister.ts:137]: to not introduce breaking change in the future, shouldn't it be called `icrcTransfer`?

Comment (peterpeterparker) [reply packages/nns/src/ledger.canister.ts:137]: or `transferIcrc`?

i.e.

- action first
- ICRC without 1

Comment (peterpeterparker) [reply packages/nns/src/ledger.canister.ts:137]: WDYT?

File: packages/nns/src/ledger.canister.ts
Comment (dskloetd) [inline packages/nns/src/ledger.canister.ts:140]: @lmuntaner:
> For HW, we will need to check the version of the Ledger App to use this method or the old one.

The check will be in the NNS-dapp, right? Not here?

File: packages/nns/src/ledger.canister.ts
Comment (dskloetd) [reply packages/nns/src/ledger.canister.ts:141]: If you review just the second commit, you'll see this is actually consistent with the existing code.
We can consider doing a big cleanup in the repo in a separate PR.

File: packages/nns/src/ledger.canister.ts
Comment (dskloetd) [reply packages/nns/src/ledger.canister.ts:137]: The canister method in the candid file is called `icrc1_transfer`.
Do you anticipate that being renamed but not changed so that we could make that breaking candid change a non-breaking change in `ic-js`?

"ICRC" can refer to a very wide variety of standards. I think it's important to point out that this is about the ICRC-1 token ledger interface standard.

File: packages/nns/src/ledger.canister.ts
Comment (dskloetd) [reply packages/nns/src/ledger.canister.ts:138]: I'm trying to stay consistent with the existing `transfer` method above.

File: packages/nns/src/ledger.canister.ts
Comment (peterpeterparker) [reply packages/nns/src/ledger.canister.ts:137]: The goal of ic-js is to offer a bridge and simple way to interact with the IC for JavaScript developer by taking care of the complexity and updates of the did files, so in that sense I would call the function `transferIcrc`.

On the other side I'm agree on the fact that there are a variety of ICRC standard. That said, I'm not sure that other chain prefix their functions with their standards' names.

Your call.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks

Can you add an entry in CHANGELOG.md before merge? Thanks.

**Pull request merged**
